spec/consensus/light-client.md

```diff
@@ -60,259 +61,368 @@ For the purpose of this lite client specification, we assume that the Tendermint
+header *h2* given the trust in the header *h1* as a single step,
+i.e., it does not assume ensuring transitive trust relation between headers through some intermediate headers.
 
-We consider the following set-up:
+In order to incentivize correct behavior of validators that run Tendermint consensus protocol, fork detection protocol (it will be explained in different document) is executed in case of a fork (conflicting
+headers are detected). As detecting conflicting headers, its propagation through the network (by the gossip protocol) and execution of the fork accountability
+protocol on the chain takes time, the lite client logic assumes conservative value for trusted period. More precisely, in the context of lite client we always
+operate with a smaller trusted period that we call *lite client trusted period* (LITE_CLIENT_TRUSTED_PERIOD). If we assume that upper bound
+for fork detection, propagation and processing on the chain is denoted with *fork procession period* (FORK_PROCESSING_PERIOD), then the following formula
+holds:
+```LITE_CLIENT_TRUSTED_PERIOD + FORK_PROCESSING_PERIOD < TRUSTED_PERIOD```, where TRUSTED_PERIOD comes from the Tendermint Failure Model.
 
-- the lite client communicates with one full node
-- the lite client locally stores all the signed headers it obtained (trusted or not). In the pseudo code below we write _Store(header)_ for this.
-- If _Bisection_ returns _false_, then the lite client has seen a forged header.
 
-  - However, it does not know which header(s) is/are the problematic one(s).
-  - In this case, the lite client can submit (some of) the headers it has seen as evidence. As the lite client communicates with one full node only when executing Bisection, there are two cases
-    - the full node is faulty
-    - the full node is correct and there was a fork in Tendermint consensus. Header _h1_ is from a different branch than the one taken by the full node. This case is not focus of this document, but will be treated in the document on fork accountability.
+*Assumption*: In the following, we assume that *h2.Header.height > h1.Header.height*. We will quickly discuss the other case in the next section.
 
-- the lite client must retry to retrieve correct headers from another full node
-  - it picks a new full node
-  - it restarts _Bisection_
-  - there might be optimizations; a lite client may not need to call _Commit(k)_, for a height _k_ for which it already has a signed header it trusts.
-  - how to make sure that a lite client can communicate with a correct full node will be the focus of a separate document (recall Issue 3 from "Context of this document").
+We consider the following set-up:
+- the lite client communicates with one full node
+- the lite client locally stores all the headers that has passed basic verification and that are within lite client trust period. In the pseudo code below we write *Store(header)* for this. If a header failed to verify, then
```

Comment:
Also, shouldn't we store header and validators in the store (header of height h, and validators of height h+1)?

Reply:
Otherwise we need to re-request validators from the RPC again in implementations. E.g. in Rust this looks like: https://github.com/interchainio/tendermint-rs/blob/27231a05fb6f12afeb220ce45c2a7f2e6b06f4e4/tendermint/src/lite/verifier.rs#L262-L266

Reply:
Yes we need to treat this better. In the spec, the headers are assumed to contain the Vals and NextVals, so we abstract over fetching the underlying data. We should probably treat it more explicitly that we need to fetch these separately (ie. two extra rpc calls to get the validator sets and check them against the hashes in the header), but maybe it can be done as a remark so that say accessing `header.Vals` or `header.NextVals`, or storing a header, does the necessary rpc calls and merkle verifications if/when necessary.

Reply:
We should be more precise when we need to make RPC as this is part of the protocol, so we need to be explicit about error conditions. I will make this explicit in the new version.